pkg/builtin/join.go

```diff
@@ -0,0 +1,134 @@
+package builtin
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/ticat/pkg/cli/core"
+)
+
+const (
+	joinKvName = "join.kvs"
+)
+
+type joinKv struct {
+	Key  string   `json:"key"`
+	Vals []string `json:"vals"`
+}
+
+func JoinNew(
+	argv core.ArgVals,
+	cc *core.Cli,
+	env *core.Env,
+	flow *core.ParsedCmds,
+	currCmdIdx int) (int, bool) {
+
+	assertNotTailMode(flow, currCmdIdx)
+
+	key := argv.GetRaw("key")
+	val := argv.GetRaw("value")
+
+	if key == "" || val == "" {
+		panic(core.NewCmdError(flow.Cmds[currCmdIdx],
+			"need key or values"))
+	}
+	vals := strings.Split(val, ",")
+	kv := joinKv{
+		Key:  key,
+		Vals: vals,
+	}
+	kvsStr := env.GetRaw(joinKvName)
+	kvs := []joinKv{}
+	if kvsStr != "" {
+		if err := json.Unmarshal([]byte(kvsStr), &kvs); err != nil {
+			panic(core.NewCmdError(flow.Cmds[currCmdIdx],
+				fmt.Sprintf("parse argument '%s' failed: %s", kvsStr, err.Error())))
+		}
+	}
+	kvs = append(kvs, kv)
+	data, err := json.Marshal(kvs)
+	if err != nil {
+		panic(core.NewCmdError(flow.Cmds[currCmdIdx],
+			fmt.Sprintf("encode argument '%s' failed: %s", kvsStr, err.Error())))
+	}
+	env.GetLayer(core.EnvLayerSession).Set(joinKvName, string(data))
+	return currCmdIdx, true
+}
+
+func JoinRun(
+	argv core.ArgVals,
+	cc *core.Cli,
+	env *core.Env,
+	flow *core.ParsedCmds,
+	currCmdIdx int) (int, bool) {
+
+	assertNotTailMode(flow, currCmdIdx)
+
+	kvs := []joinKv{}
+	kvsStr := env.GetRaw(joinKvName)
+	if err := json.Unmarshal([]byte(kvsStr), &kvs); err != nil {
+		panic(core.NewCmdError(flow.Cmds[currCmdIdx],
+			fmt.Sprintf("parse argument '%s' failed: %s", kvsStr, err.Error())))
+	}
+	cmd := argv.GetRaw("cmd")
+	if cmd == "" {
+		panic(core.NewCmdError(flow.Cmds[currCmdIdx],
+			"can't execute null ticat command"))
+	}
+	iter := newJoinKvsIter(kvs)
+	for {
+		args := []string{cmd}
+		keys, vals := iter.get()
+		for i, key := range keys {
+			args = append(args, key+"="+vals[i])
+		}
+		cc.Executor.Execute("join", true, cc, env, nil, args...)
```

Comment:
use `flow.Cmds[currCmdIdx].DisplayPath(cc.Cmds.Strs.PathSep, false)` instead of "join", because we don't know for sure this command will be register as command "join" or other names